File: Day18/main.cpp
#include<array>
#include<chrono>
#include<iostream>
#include<fstream>
#include<sstream>
#include<string>
#include<utility>
#include<vector>
#ifdef _MSC_VER
#define UNREACHABLE() __assume(false)
#define ASSUME(C) __assume(C)
#else
#define UNREACHABLE() __builtin_unreachable()
#define ASSUME(C) do{if(!__builtin_expect(C,true)){UNREACHABLE();}}while(false)
#endif
using t_t=decltype(std::chrono::high_resolution_clock::now());
t_t getTime(){
   return std::chrono::high_resolution_clock::now();
}
using d_t=std::chrono::nanoseconds;
double getDurationNs(t_t start,t_t end){
   return std::chrono::duration_cast<d_t>(end-start).count()/1.0;
}
double getDurationUs(t_t start,t_t end){
   return std::chrono::duration_cast<d_t>(end-start).count()/1'000.0;
}
double getDurationMs(t_t start,t_t end){
   return std::chrono::duration_cast<d_t>(end-start).count()/1'000'000.0;
}
double getDurationS(t_t start,t_t end){
   return std::chrono::duration_cast<d_t>(end-start).count()/1'000'000'000.0;
}
using shint=unsigned char;
using llint=long long int;
using Line=std::string;
using Input=std::vector<std::string>;
struct Position{
   llint x;
   llint y;
   bool isIn(llint width,llint height)const{
      return 0<=x&&x<width&&0<=y&&y<height;
   }
   template<typename T>bool isIn(T &t)const{
      return isIn(t[0].size(),t.size());
   }
   template<typename T>decltype(std::declval<T>()[0][0])indexInto(T &t)const{
      return t[y][x];
   }
   friend auto operator<=>(Position l,Position r){
      if(auto dx=l.x<=>r.x;dx!=0){
         return dx;
      }
      return l.y<=>r.y;
   }
};
using Cell=bool;
using Row=std::vector<Cell>;
using Board=std::vector<Row>;
#include<ios>
llint solve1(const Input &input){
   constexpr llint BOARDSIZE=300;
   Board board(2*BOARDSIZE,Row(2*BOARDSIZE,false));
   Position position{BOARDSIZE,BOARDSIZE};
   for(const Line &_line:input){
      std::istringstream line{_line};
      shint direction;
      line>>direction;
      line.ignore(1);
      llint length;
      line>>length;
      while(length--){
         position.indexInto(board)=true;
         switch(direction){
         case 'U':position.y--;break;
         case 'R':position.x++;break;
         case 'D':position.y++;break;
         case 'L':position.x--;break;
         }
      }
   }
   llint count=0;
   for(position.y=0;position.y<2*BOARDSIZE;position.y++){
      bool isInside=false;
      bool cameDown=false;
      bool cameUp=false;
      for(position.x=0;position.x<2*BOARDSIZE;position.x++){
         if(position.indexInto(board)){
            bool goingUp=0<=position.y-1&&board[position.y-1][position.x];
            bool goingDown=position.y+1<2*BOARDSIZE&&board[position.y+1][position.x];
            if(goingUp&&goingDown){
               isInside=!isInside;
            } else if(!cameDown&&!cameUp){
               cameDown=goingUp;
               cameUp=goingDown;
            } else if((cameDown&&goingUp)||(cameUp&&goingDown)){
               cameDown=false;
               cameUp=false;
            } else if((cameDown&&goingDown)||(cameUp&&goingUp)){
               isInside=!isInside;
               cameDown=false;
               cameUp=false;
            }
            count++;
         } else{
            if(isInside){
               count++;
            }
         }
      }
   }
   return count;
}
llint solve2(const Input &input){
   std::vector<Position>corners;
   Position position(0,0);
   corners.push_back(position);
   llint borderCount=0;
   for(const Line &line:input){
      -
         std::istringstream line{_line};
      shint direction;
      line>>direction;
      line.ignore(1);
      llint length;
      line>>length;
      line.ignore(3);
      llint color;
      line>>std::hex>>color;
      length=color/16;
      switch(color%16){
      case 0:position.x+=length;break;
      case 1:position.y+=length;break;
      case 2:position.x-=length;break;
      case 3:position.y-=length;break;
      }
      corners.push_back(position);
      borderCount+=length;
   }
   llint count=0;
   for(llint i=0;i<corners.size()-1;i++){
      Position start=corners[i];
      Position end=corners[i+1];
      llint cross=start.x*end.y-end.x*start.y;
      count+=cross;
   }
   return std::abs(count/2)+borderCount/2+1;
}
int main(){
   constexpr bool logToFile=false;
   constexpr llint doWarming=100000;
   constexpr bool doExample1=true;
   constexpr bool doInput1=true;
   constexpr bool doExample2=true;
   constexpr bool doInput2=true;

   Input example1={
      "R 6 (#70c710)",
      "D 5 (#0dc571)",
      "L 2 (#5713f0)",
      "D 2 (#d2c081)",
      "R 2 (#59c680)",
      "D 2 (#411b91)",
      "L 5 (#8ceee2)",
      "U 2 (#caa173)",
      "L 1 (#1b58a2)",
      "U 2 (#caa171)",
      "R 2 (#7807d2)",
      "U 3 (#a77fa3)",
      "L 2 (#015232)",
      "U 2 (#7a21e3)"
   };

   Input example2=example1;

   Input input;
   std::ifstream inputFile("input");
   Line line;
   while(std::getline(inputFile,line)){
      input.push_back(line);
   }
   inputFile.close();

   volatile llint result=0;
   for(llint warming=0;warming<doWarming;warming++){
      result=(doExample1?solve1(example1):-1)+(doInput1?solve1(input):-1)
         +(doExample2?solve2(example2):-1)+(doInput2?solve2(input):-1);
   }

   t_t time0,time1,time2,time3,time4;
   decltype(solve1(example1))answer1example=-1;
   decltype(solve1(input))answer1input=-1;
   decltype(solve1(example2))answer2example=-1;
   decltype(solve1(input))answer2input=-1;

   std::ofstream log;
   std::streambuf *coutBuf;
   if constexpr(logToFile){
      log.open("log.txt");
      coutBuf=std::cout.rdbuf();
      std::cout.rdbuf(log.rdbuf());
   }

   time0=getTime();
   if constexpr(doExample1){
      answer1example=solve1(example1);
   }
   time1=getTime();
   if constexpr(doInput1){
      answer1input=solve1(input);
   }
   time2=getTime();
   if constexpr(doExample2){
      answer2example=solve2(example2);
   }
   time3=getTime();
   if constexpr(doInput2){
      answer2input=solve2(input);
   }
   time4=getTime();

   if constexpr(logToFile){
      std::cout.rdbuf(coutBuf);
      log.close();
   }

   std::cout<<"Example 1 took "<<getDurationMs(time0,time1)<<"ms: "<<answer1example<<'\n';
   std::cout<<"Input   1 took "<<getDurationMs(time1,time2)<<"ms: "<<(answer1input)<<'\n';
   std::cout<<"Example 2 took "<<getDurationMs(time2,time3)<<"ms: "<<answer2example<<'\n';
   std::cout<<"Input   2 took "<<getDurationMs(time3,time4)<<"ms: "<<(answer2input)<<'\n';
}

File: Day18/main.py
import time
from heapq import heappush,heappop

def solve(array,start,end):
   Y=len(array)
   X=len(array[0])
   heap=[(0,1,start,[])]
   done=set()
   while heap:
      s,d,p,h=heappop(heap)
      x,y=p
      if not(0<=x and x<X and 0<=y and y<Y):
         pass
      elif p==end:
         h=h+[p]
         return h
      elif array[y][x] and p not in done:
         done.add(p)
         h=h+[p]
         heappush(heap,(s+1,0,(x,y-1),h))
         heappush(heap,(s+1,1,(x+1,y),h))
         heappush(heap,(s+1,2,(x,y+1),h))
         heappush(heap,(s+1,3,(x-1,y),h))
   return []

def solve1(cs,X,Y,n):
   a=[[True for x in range(X)] for y in range(Y)]
   for i in range(n):
      c=cs[i]
      a[c[1]][c[0]]=False
   return len(solve(a,(0,0),(X-1,Y-1)))-1

def solve2(cs,X,Y,n):
   a=[[True for x in range(X)] for y in range(Y)]
   for i in range(n):
      c=cs[i]
      a[c[1]][c[0]]=False
   start=(0,0)
   end=(X-1,Y-1)
   prev=[]
   for i in range(n,len(cs)):
      c=cs[i]
      a[c[1]][c[0]]=False
      if not prev or c in prev:
         cur=solve(a,start,end)
         if not cur:
            return f'{c[0]},{c[1]}'
         prev=cur
   return '-1,-1'

def readsplit(filename):
   with open(filename,'r') as file:
      return file.read().rstrip('\n').split('\n')

def main(filename='input'):
   time0=time.perf_counter_ns()

   input=readsplit(filename)

   time1=time.perf_counter_ns()

   coordinates=[]
   for line in input:
      coordinates.append(tuple(int(i) for i in line.split(',')))
   Y=71
   X=71
   n=1024
   if filename=='example':
      Y=7
      X=7
      n=12

   time2=time.perf_counter_ns()

   answer1=solve1(coordinates,X,Y,n)

   time3=time.perf_counter_ns()

   answer2=solve2(coordinates,X,Y,n)

   time4=time.perf_counter_ns()

   read=(time1-time0)/1000000.0
   parse=(time2-time1)/1000000.0
   part1=(time3-time2)/1000000.0
   part2=(time4-time3)/1000000.0
   print(f'Read     input  in {read}ms')
   print(f'Parsed   input  in {parse}ms')
   print(f'Answered part 1 in {part1}ms: {answer1}')
   print(f'Answered part 2 in {part2}ms: {answer2}')

   return ((answer1,answer2),(read,parse,part1,part2))

if __name__=='__main__':
   main('example')
   main()
